graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/fetchers/admin/CreateNamespaceFetcher.java

```diff
@@ -0,0 +1,131 @@
+import io.stargate.db.datastore.DataStore;
+import io.stargate.db.datastore.DataStoreFactory;
+import io.stargate.db.query.builder.Replication;
+import io.stargate.db.schema.Keyspace;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import org.apache.cassandra.stargate.exceptions.AlreadyExistsException;
+
+public class CreateNamespaceFetcher
+    extends NamespaceFetcher<DataFetcherResult<Map<String, Object>>> {
+
+  public CreateNamespaceFetcher(
+      AuthorizationService authorizationService, DataStoreFactory dataStoreFactory) {
+    super(authorizationService, dataStoreFactory);
+  }
+
+  @Override
+  protected DataFetcherResult<Map<String, Object>> get(
+      DataFetchingEnvironment environment,
+      DataStore dataStore,
+      AuthenticationSubject authenticationSubject) {
+
+    String name = environment.getArgument("name");
+
+    try {
+      authorizationService.authorizeSchemaWrite(
+          authenticationSubject, name, null, Scope.CREATE, SourceAPI.GRAPHQL);
+    } catch (UnauthorizedException e) {
+      return errorResult(e, environment).build();
+    }
+
+    boolean ifNotExists = environment.getArgumentOrDefault("ifNotExists", Boolean.FALSE);
+    Integer replicas = environment.getArgument("replicas");
+    List<Map<String, Object>> datacenters = environment.getArgument("datacenters");
+    if (replicas == null && datacenters == null) {
+      replicas = 1;
+    }
+    if (replicas != null && datacenters != null) {
+      throw new IllegalArgumentException("You can't specify both replicas and datacenters");
+    }
+    Replication replication =
+        replicas != null
+            ? Replication.simpleStrategy(replicas)
+            : Replication.networkTopologyStrategy(parseDatacenters(datacenters));
+
+    DataFetcherResult.Builder<Map<String, Object>> result;
+    try {
+      dataStore
+          .execute(
+              dataStore
+                  .queryBuilder()
+                  .create()
+                  .keyspace(name)
+                  .ifNotExists(ifNotExists)
+                  .withReplication(replication)
+                  .build()
+                  .bind())
+          .get();
+      result = DataFetcherResult.newResult();
+    } catch (Exception e) {
+      if (e instanceof ExecutionException && e.getCause() instanceof AlreadyExistsException) {
+        // Error out, but we'll still include the details of the existing keyspace below.
+        result = errorResult(e.getCause(), environment);
+      } else {
+        // Unexpected technical error: rethrow directly
+        return errorResult(e, environment).build();
+      }
+    }
+
+    ImmutableMap.Builder<String, Object> data = ImmutableMap.builder();
+    data.put("query", environment.getGraphQLSchema().getQueryType());
+
+    Keyspace keyspace = dataStore.schema().keyspace(name);
+    if (isReadAuthorized(keyspace, authenticationSubject)) {
```

Comment:
Is this necessary since we're essentially returning what was sent in?

Reply:
I was thinking of the case where `ifNotExists == true`, and the keyspace already existed.